README.md
# about
dead simple crontab powered job scheduling for django.

# setup
install via easy_install or pip

        easy_install django-crontab

add it to installed apps in django settings.py

        INSTALLED_APPS = (
            'django_crontab',
            ...
        )

now create a new method that should be executed by cron every 5 minutes, f.e. in `myproject/myapp/cron.py`

        def my_scheduled_job():
          pass

now add this to your settings.py:

        CRONJOBS = [
            ('*/5 * * * *', 'myproject.myapp.cron.my_scheduled_job')
        ]

the least to do is to run this command to add all defined jobs from `CRONJOBS` to crontab (of the user which you are running this command with):

        python manage.py crontab add

removing all defined jobs is straight forward

        python manage.py crontab remove

# config
there are a bunch of setting vars to customize behavior. each of this comes with default values that should properly fit. if not, feel free to overwrite.

* CRONJOBS
  * list of tuples with cron timing, the python module path to the method and optional a job specific suffix (f.e. to redirect out/err to a file)
  * default: []
  * example
        
            CRONJOBS = [
                ('*/5 * * * *', 'myproject.myapp.cron.my_scheduled_job'),
                ('0   0 1 * *', 'myproject.myapp.cron.other_scheduled_job'),
                ('@reboot',     'myproject.anotherapp.cron.system_reboot_job', '>> /home/john/web/logs/system_reboot_job.log'),
            ]

* CRONTAB\_EXECUTABLE
  * path to the crontab executable of your os
  * default: '/usr/bin/crontab'

* CRONTAB\_DJANGO\_PROJECT\_NAME
  * the name of your django project, used to build path path to manage.py and to mark the jobs in contrab via comment for later removing
  * default is read from DJANGO_SETTINGS_MODULE environment variable 

* CRONTAB\_DJANGO\_MANAGE\_PATH
  * path to manage.py file (including the manage.py itself, i.e. '/home/john/web/manage.py')
  * default is build using DJANGO\_PROJECT\_NAME

* CRONTAB\_PYTHON\_EXECUTABLE
  * path to the python interpreter executable used to run the scheduled job
  * default uses the interpreter executable used to `add` the jobs (via 'python manage.py crontab add')

* CRONTAB\_COMMAND\_PREFIX
  * something you wanne do before job gets executed.
  * default: '' (empty string) 
  * example: '`echo "executing my scheduled job now" && `'

* CRONTAB\_COMMAND\_SUFFIX
  * something you wanne do after job was executed.
  * default: '' (empty string) 
  * example: '` && echo "execution of my scheduled job finished"`'

* CRONTAB\_COMMENT
  * used for marking the added contab-lines for removing, default value includes project name to distinguish multiple projects on the same host and user
  * default: 'django-crontabs for ' + CRONTAB\_DJANGO\_PROJECT\_NAME

# license
MIT-License, see [LICENSE](/kraiz/django-crontab/blob/master/LICENSE) file.


CHANGELOG
0.7.1
-----
* bugfix: allow project path to contain "run" #30 (thx vinsia)

0.7.0
------
* dropped Django 1.7 and below, you'll need at least 1.8
* added support for Django 1.9 + Python 3.5

0.6.0
------
* refactoring for being better testing (big thx arski)
* feature: TESTS! > 90% coverage
* feature: compatiblity with Django 1.3 - 1.7 and Python 3.3 + 3.4
* feature: ability to run the scheduled command with a different settings module #10 (thx chamaken)
* bugfix: fixed unsupported compare between str and int in python 3.4 #19 (thx justdoit0823)

0.5.1
------
* bugfix: added 2 formats for jobs to keep backward compatiblity with 0.4.x
* feature: subcommand "show" that lists currently active jobs from crontab

0.5.0
------
* feature: positional and keyword arguments for job functions as third and fourth job tuple element. this enables django
           management commands to be called, see documentation
           NOTE: this is a breaking change when you used job suffixes, they will move from third
                 to fifth job tuple element
* feature: prevent starting a job if an old instance of the same job is still running (thx arski)
* feature: logging job's possible exceptions (thx arski)

0.4.3
------
* bugfix: only replace "pyc" with "py" in file's extension (thx cinghiale)

0.4.2
------
* changed README from markdown to restructuredtext
* fixed a crappy distribution error with README

0.4.1 - broken!
------
* bugfix: list modification while iterating d77a48a (thx meric426)
* feature: setting CRONTAB_COMMENT overwritable now

0.4
------
* bugfix: crash related to Django 1.4 and move of manage.py

0.3
------
* new feature: added support for command pre- and suffixes (f.e. pipe stout/stderr into logfile)

0.2
------
* first public release


tests/cron.py


def cron_job():
    pass


tests/__init__.py
import os

from django.test.utils import setup_test_environment


os.environ.setdefault("DJANGO_SETTINGS_MODULE", "settings")
setup_test_environment()


tests/settings.py
# Django settings for test_project project.
DEBUG = True

INSTALLED_APPS = (
    'django_crontab',
)

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3'
    }
}

MIDDLEWARE_CLASSES = ()

SECRET_KEY = 'not-so-secret'

LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'filters': {},
    'handlers': {
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler'
        }
    },
    'loggers': {
        'django_crontab': {
            'handlers': ['console'],
            'level': 'INFO',
        }
    }
}

import os
CRONTAB_DJANGO_MANAGE_PATH = os.path.join(os.path.dirname(__file__), '', 'test_manage.py')


django_crontab/__init__.py
from __future__ import print_function

import fcntl
import hashlib
import json
import logging
import os
import tempfile

from django.conf import settings

from django.utils.importlib import import_module

from django_crontab.app_settings import Settings

logger = logging.getLogger(__name__)


class Crontab(object):

    def __init__(self, **options):
        self.verbosity = options.get('verbosity', 1)
        self.readonly = options.get('readonly', False)
        self.crontab_lines = []
        self.settings = Settings(settings)

    def __enter__(self):
        """
        Automatically read crontab when used as with statement
        """
        self.read()
        return self

    def __exit__(self, type, value, traceback):
        """
        Automatically write back crontab when used as with statement if readonly is False
        """
        if not self.readonly:
            self.write()

    def read(self):
        """
        Reads the crontab into internal buffer
        """
        self.crontab_lines = os.popen('%s -l' % self.settings.CRONTAB_EXECUTABLE).readlines()

    def write(self):
        """
        Writes internal buffer back to crontab
        """
        fd, path = tempfile.mkstemp()
        tmp = os.fdopen(fd, 'w')
        for line in self.crontab_lines:
            tmp.write(line)
        tmp.close()
        os.system('%s %s' % (self.settings.CRONTAB_EXECUTABLE, path))
        os.unlink(path)

    def add_jobs(self):
        """
        Adds all jobs defined in CRONJOBS setting to internal buffer
        """
        print(self.settings.CRONJOBS)
        for job in self.settings.CRONJOBS:
            # differ format and find job's suffix
            if len(job) > 2 and isinstance(job[2], (basestring, unicode)):
                # format 1 job
                job_suffix = job[2]
            elif len(job) > 4:
                job_suffix = job[4]
            else:
                job_suffix = ''

            self.crontab_lines.append(self.settings.CRONTAB_LINE_PATTERN % {
                'time': job[0],
                'comment': self.settings.CRONTAB_COMMENT,
                'command': (
                    '%(global_prefix)s %(exec)s %(manage)s crontab run '
                    '%(jobname)s %(job_suffix)s %(global_suffix)s'
                ) % {
                    'global_prefix': self.settings.COMMAND_PREFIX,
                    'exec': self.settings.PYTHON_EXECUTABLE,
                    'manage': self.settings.DJANGO_MANAGE_PATH,
                    'jobname': self.__hash_job(job),
                    'job_suffix': job_suffix,
                    'global_suffix': self.settings.COMMAND_SUFFIX
                }
            })
            if self.verbosity >= 1:
                print('  adding cronjob: (%s) -> %s' % (self.__hash_job(job), job))

    def show_jobs(self):
        """
        Print the jobs from from crontab
        """
        print("Currently active jobs in crontab:")
        for line in self.crontab_lines[:]:
            job = self.settings.CRONTAB_LINE_REGEXP.findall(line)
            if job and job[0][4] == self.settings.CRONTAB_COMMENT:
                if self.verbosity >= 1:
                    print('%s -> %s' % (
                        job[0][2].split()[4],
                        self.__get_job_by_hash(job[0][2][job[0][2].find('run') + 4:].split()[0])
                    ))

    def remove_jobs(self):
        """
        Removes all jobs defined in CRONJOBS setting from internal buffer
        """
        for line in self.crontab_lines[:]:
            job = self.settings.CRONTAB_LINE_REGEXP.findall(line)
            if job and job[0][4] == self.settings.CRONTAB_COMMENT:
                self.crontab_lines.remove(line)
                if self.verbosity >= 1:
                    print('removing cronjob: (%s) -> %s' % (
                        job[0][2].split()[4],
                        self.__get_job_by_hash(job[0][2][job[0][2].find('run') + 4:].split()[0])
                    ))

    def __hash_job(self, job):
        """
        Builds an md5 hash representing the job
        """
        return hashlib.md5(json.dumps(job)).hexdigest()

    def get_job_by_hash(self, job_hash):
        """
        Finds the job by given hash
        """
        for job in self.settings.CRONJOBS:
            if self.__hash_job(job) == job_hash:
                return job
        return None

    # noinspection PyBroadException
    def __run_cronjob(self, job_hash):
        """
        Executes the corresponding function defined in CRONJOBS
        """
        job = self.__get_job_by_hash(job_hash)
        job_name = job[1]
        job_args = job[2] if len(job) > 2 else []
        job_kwargs = job[3] if len(job) > 3 else {}

        lock_file_name = None
        if self.settings.LOCK_JOBS:
            lock_file = open(os.path.join(tempfile.gettempdir(), 'django_crontab_%s.lock' % job_hash), 'w')
            lock_file_name = lock_file.name
            try:
                fcntl.lockf(lock_file, fcntl.LOCK_EX | fcntl.LOCK_NB)
            except:
                logger.warning('Tried to start cron job %s that is already running.', job)
                return

        module_path, function_name = job_name.rsplit('.', 1)
        module = import_module(module_path)
        func = getattr(module, function_name)
        try:
            func(*job_args, **job_kwargs)
        except:
            logger.exception('Failed to complete cronjob at %s', job)

        if self.settings.LOCK_JOBS:
            try:
                os.remove(lock_file_name)
            except:
                logger.exception('Error deleting lockfile %s of job %s', lock_file_name, job)


django_crontab/management/commands/crontab.py
from __future__ import print_function

from django.core.management.base import BaseCommand
from django_crontab.crontab import Crontab


class Command(BaseCommand):
    args = '<add|show|remove>'
    help = 'run this command to add, show or remove the jobs defined in CRONJOBS setting from/to crontab'
    crontab_lines = []

    def add_arguments(self, parser):
        parser.add_argument('subcommand', help='subcommand to run, see help')
        parser.add_argument('jobhash', nargs='?', default='')

    def handle(self, *args, **options):
        """
        Dispatches by given subcommand
        """
        if len(args) > 0:
            if args[0] == 'add':
                with Crontab(**options) as crontab:
                    crontab.remove_jobs()
                    crontab.add_jobs()
                return
            elif args[0] == 'show':
                with Crontab(readonly=True, **options) as crontab:
                    crontab.show_jobs()
                return
            elif args[0] == 'remove':
                with Crontab(**options) as crontab:
                    crontab.remove_jobs()
                return
            elif args[0] == 'run':
                if len(args) > 1:
                    jobhash = args[1]
                else:
                    jobhash = options['jobhash']

                Crontab().run_job(jobhash)
                return
        print(help)


django_cronjobs/management/commands/cronjobs.py
from django.core.management.base import BaseCommand
from django.utils.importlib import import_module
from django_cronjobs.app_settings import CRONTAB_EXECUTABLE, CRONJOBS, \
    CRONTAB_LINE_PATTERN, CRONTAB_COMMENT, PYTHON_EXECUTABLE, DJANGO_MANAGE_PATH, \
    CRONTAB_LINE_REGEXP, COMMAND_PREFIX, COMMAND_SUFFIX
import os
import tempfile

class Command(BaseCommand):
    args = '<add|remove>'
    help = 'run this command to add or remove the jobs defined in CRONJOBS setting from/to crontab'
    crontab_lines = []


    def handle(self, *args, **options):
        """dispatches by given subcommand"""
        if len(args) > 0:
            if args[0] == 'add':
                self.__read_crontab(**options)
                self.__remove_cronjobs(**options)
                self.__add_cronjobs(**options)
                self.__write_crontab(**options)
                return
            elif args[0] == 'remove':
                self.__read_crontab(**options)
                self.__remove_cronjobs(**options)
                self.__write_crontab(**options)
                return
            elif args[0] == 'run':
                self.__run_cronjob(args[1])
                return
        print help


    def __read_crontab(self, **options):
        """reads the crontab into internal buffer"""
        if options.get('verbosity') == '2':
            print 'reading from crontab...',
        self.crontab_lines = os.popen('%s -l' % CRONTAB_EXECUTABLE).readlines()
        if options.get('verbosity') == '2':
            print 'done'


    def __write_crontab(self, **options):
        """writes internal buffer back to crontab"""
        if options.get('verbosity') == '2':
            print 'writing to crontab...',
        fd, path = tempfile.mkstemp()
        tmp = os.fdopen(fd, 'w')
        for line in self.crontab_lines:
            tmp.write(line)
        tmp.close()
        os.system('%s %s' % (CRONTAB_EXECUTABLE, path))
        os.unlink(path)
        if options.get('verbosity') == '2':
            print 'done'


    def __add_cronjobs(self, **options):
        """adds all jobs defined in CRONJOBS setting to internal buffer"""
        for cronjob in CRONJOBS:
            self.crontab_lines.append(CRONTAB_LINE_PATTERN % {
                'time': cronjob[0],
                'comment': CRONTAB_COMMENT,
                'command': '%(prefix)s %(exec)s %(manage)s cronjobs run %(jobname)s %(suffix)s' % {
                    'prefix': COMMAND_PREFIX,
                    'exec': PYTHON_EXECUTABLE,
                    'manage': DJANGO_MANAGE_PATH,
                    'jobname': cronjob[1],
                    'suffix': COMMAND_SUFFIX
                }
            })
            if options.get('verbosity') == '1':
                print '  adding cronjob: %s -> %s' % cronjob
            elif options.get('verbosity') == '2':
                print 'adding cronjob: %s' % self.crontab_lines[-1],


    def __remove_cronjobs(self, **options):
        """removes all jobs defined in CRONJOBS setting from internal buffer"""
        for line in self.crontab_lines:
            job = CRONTAB_LINE_REGEXP.findall(line)
            if job and job[0][4] == CRONTAB_COMMENT:
                self.crontab_lines.remove(line)
                if options.get('verbosity') == '1':
                    print 'removing cronjob: %s -> %s' % (job[0][0].strip(), job[0][2][job[0][2].find('run') + 4:].split()[0])
                elif options.get('verbosity') == '2':
                    print 'removing cronjob: %s ' % line,


    def __run_cronjob(self, function):
        """executes the corresponding function defined in CRONJOBS"""
        for cronjob in CRONJOBS:
            if cronjob[1] == function:
                module_path = function.split('.')[:-1][0]
                module = import_module(module_path)
                func = getattr(module, function.split('.')[-1])
                func()



django_crontab/app_settings.py
from __future__ import print_function

import os
import re
import sys

from django.utils.importlib import import_module


class Settings():
    def __init__(self, settings):
        self.CRONJOBS = getattr(settings, 'CRONJOBS', [])

        self.CRONTAB_EXECUTABLE = getattr(settings, 'CRONTAB_EXECUTABLE', '/usr/bin/crontab')

        self.CRONTAB_LINE_REGEXP = re.compile(r'^\s*(([^#\s]+\s+){5})([^#\n]*)\s*(#\s*([^\n]*)|$)')

        self.CRONTAB_LINE_PATTERN = '%(time)s %(command)s # %(comment)s\n'

        self.DJANGO_PROJECT_NAME = getattr(settings, 'CRONTAB_DJANGO_PROJECT_NAME', os.environ['DJANGO_SETTINGS_MODULE'].split('.')[0])

        self.DJANGO_SETTINGS_MODULE = getattr(settings, 'CRONTAB_DJANGO_SETTINGS_MODULE', None)

        if hasattr(settings, 'CRONTAB_DJANGO_MANAGE_PATH'):
            self. DJANGO_MANAGE_PATH = settings.CRONTAB_DJANGO_MANAGE_PATH
            # check if it's really there
            if not os.path.exists(self.DJANGO_MANAGE_PATH):
                print('ERROR: No manage.py file found at "%s". Check settings.CRONTAB_DJANGO_MANAGE_PATH!' % self.DJANGO_MANAGE_PATH)
        else:
            def ext(fpath):
                return os.path.splitext(fpath)[0] + '.py'
            try:  # Django 1.3
                self.DJANGO_MANAGE_PATH = ext(import_module(self.DJANGO_PROJECT_NAME + '.manage').__file__)
            except ImportError:
                try:  # Django 1.4+
                    self.DJANGO_MANAGE_PATH = ext(import_module('manage').__file__)
                except ImportError:
                    print('ERROR: Can\'t find your manage.py - please define settings.CRONTAB_DJANGO_MANAGE_PATH')

        self.PYTHON_EXECUTABLE = getattr(settings, 'CRONTAB_PYTHON_EXECUTABLE', sys.executable)

        self.CRONTAB_COMMENT = getattr(settings, 'CRONTAB_COMMENT', 'django-cronjobs for %s' % self.DJANGO_PROJECT_NAME)

        self.COMMAND_PREFIX = getattr(settings, 'CRONTAB_COMMAND_PREFIX', '')
        self.COMMAND_SUFFIX = getattr(settings, 'CRONTAB_COMMAND_SUFFIX', '')

        self.LOCK_JOBS = getattr(settings, 'CRONTAB_LOCK_JOBS', False)
